### PR TITLE
Port x-lang support to kinesis v2

### PR DESCRIPTION
**Pull request by xinbinhuang**
**Please** add a meaningful description for your change here

cc: @aromanenko-dev @mosche 

fixes: #23570

I've implemented the Java part and have some questions wanna clear before adding the Python part:

Java Questions:
- do I need to add the expand service gradle config: https://github.com/apache/beam/tree/master/sdks/java/io/kinesis/expansion-service?
- What would be the proper input type to `KinesisIO.Write`? I'm not sure how Python type translate to x-lang transforms.. I hard-coded it to `byte[]` for now.
- Why does `KinesisIO` implements its own `KinesisIO.Write.Result` instead of using the provided `PDone`? They looks the same in terms of implementation.
- Testing: is adding tests to [xlang_kinesisio_it_test.py](https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/external/xlang_kinesisio_it_test.py) after changes on the python part the correct approach?

Python questions:
- Because some fields are removed (i.e. `partition_key`, `verify_certificate`), and mechanisms are changes (i.e. partitioning). What's the general policy for beam to introduce breaking changes? Do we just break it at the code-level and release it with major version? Or do we introduce a v2 version and deprecate the old one to allow gradual transition.  

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by aromanenko-dev** [comment]
@xinbinhuang Many thanks for working on this!

Ping @chamikaramj to help with x-lang questions

File: sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisTransformRegistrar.java [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+      }
+    }
+
+    private enum WatermarkPolicy {
+      ARRIVAL_TIME,
+      PROCESSING_TIME
+    }
+
+    @Override
+    public PTransform<PBegin, PCollection<KinesisRecord>> buildExternal(
```
Comment (mosche) [inline sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/KinesisTransformRegistrar.java:228]: The output needs to be `byte[]`. You can just add a `MapElements` to extract `data` from the record.

**Review by mosche — COMMENTED** [review]
Thanks @xinbinhuang !

@chamikaramj can you provide more details? Pls correct me if I'm wrong below.

> do I need to add the expand service gradle config: https://github.com/apache/beam/tree/master/sdks/java/io/kinesis/expansion-service?

Yes, the expansion service is needed to run the Java side.

> What would be the proper input type to KinesisIO.Write? I'm not sure how Python type translate to x-lang transforms. I hard-coded it to byte[] for now.

`byte[]` seems ok for both reading and writing. Though, that makes it hard to use decent, content based partitioning (see comment on random partitioning). For writing a tuple type such as `KV` would make a lot more sense. That allows users to pass a partition key. Have a look at the external [KafkaIO](https://github.com/apache/beam/blob/master/sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java#L2756) as example.

Similarly, the record metadata might be required for certain use cases when reading. As far as I know hat requires a schema for KinesisRecord, but I'm not sure of the contracts / how this is supposed to be done. In any case, just returning the binary data seems to be ok in the read case.

> Why does KinesisIO implements its own KinesisIO.Write.Result instead of using the provided PDone? They looks the same in terms of implementation.

This is a [Beam convention](https://beam.apache.org/documentation/io/io-standards/#general) improve evolution of write APIs:

    The top-level I/O class will provide a static method to start constructing an I/O.Write 
    transform. This returns a PTransform with a single input PCollection, and a Write.Result 
    output.

For instance this allows to add more advanced error handling, e.g. adding a mode that returns failed records rather than failing fast.

> Because some fields are removed (i.e. partition_key, verify_certificate), and mechanisms are changes (i.e. partitioning). What's the general policy for beam to introduce breaking changes? Do we just break it at the code-level and release it with major version? Or do we introduce a v2 version and deprecate the old one to allow gradual transition.

This is a tough one... I suppose it would be possible to silently switch the implementation on the Java side carefully setting defaults and ignoring obsolete settings (such as the old partitioning key) with appropriate warnings added. On the other hand, if types are also changed it's impossible to not break anything. That would require to add a v2, deprecate the previous one and remove it after 3+ releases.

**Comment by mosche** [comment]
Kind ping @xinbinhuang

**Comment by aromanenko-dev** [comment]
Hi @xinbinhuang, kind ping on this

**Pull request closed without merging**